File: DrowsinessApp/lib/services/detector.dart
import 'dart:io';
import 'dart:typed_data';

import 'package:camera/camera.dart';
import 'package:flutter/services.dart' show rootBundle;
import 'package:opencv_dart/opencv_dart.dart' as cv;
import 'package:path/path.dart' as p;
import 'package:path_provider/path_provider.dart';
import 'package:tflite_flutter/tflite_flutter.dart';

import '../models/types.dart';

/// Loads the TFLite model + Haar cascades, and runs detection on a frame.
///
/// Class index map (from notebook):
/// 0=yawn, 1=no_yawn, 2=Closed, 3=Open, 4=front, 5=down
class Detector {
  static const int imgSize = 224;
  static const _faceIndices = [0, 1, 4, 5];
  static const _eyeIndices = [2, 3];
  static const _faceLabels = [
    FaceClass.yawn,
    FaceClass.noYawn,
    FaceClass.front,
    FaceClass.down,
  ];

  Interpreter? _interp;
  cv.CascadeClassifier? _faceCascade;
  cv.CascadeClassifier? _eyeCascade;

  bool get isReady =>
      _interp != null && _faceCascade != null && _eyeCascade != null;

  Future<void> init({void Function(String)? onProgress}) async {
    onProgress?.call('Loading model…');
    _interp = await Interpreter.fromAsset(
      'assets/models/drowsiness_efficientnet_b0.tflite',
    );
    _interp!.allocateTensors();

    onProgress?.call('Loading cascades…');
    final faceXml = await _writeAssetToTemp(
        'assets/haarcascades/haarcascade_frontalface_default.xml',
        'face.xml');
    final eyeXml = await _writeAssetToTemp(
        'assets/haarcascades/haarcascade_eye.xml', 'eye.xml');
    _faceCascade = cv.CascadeClassifier.fromFile(faceXml);
    _eyeCascade = cv.CascadeClassifier.fromFile(eyeXml);

    onProgress?.call('Ready');
  }

  Future<String> _writeAssetToTemp(String asset, String name) async {
    final dir = await getTemporaryDirectory();
    final path = p.join(dir.path, name);
    final f = File(path);
    if (!await f.exists()) {
      final bytes = await rootBundle.load(asset);
      await f.writeAsBytes(bytes.buffer.asUint8List());
    }
    return path;
  }

  /// Run detection on a CameraImage frame (mobile path).
  DetectionResult detect(CameraImage image, double confThreshold) {
    if (!isReady) {
      return DetectionResult(
        faces: const [],
        frameWidth: image.width,
        frameHeight: image.height,
        tsMs: DateTime.now().millisecondsSinceEpoch,
      );
    }
    final mat = _matFromCameraImage(image);
    if (mat == null) {
      return DetectionResult(
        faces: const [],
        frameWidth: image.width,
        frameHeight: image.height,
        tsMs: DateTime.now().millisecondsSinceEpoch,
      );
    }
    try {
      return detectMat(mat, confThreshold);
    } finally {
      mat.dispose();
    }
  }

  /// Run detection on an already-decoded BGR cv.Mat (desktop path).
  /// Caller owns the Mat lifecycle.
  DetectionResult detectMat(cv.Mat mat, double confThreshold) {
    if (!isReady) {
      return DetectionResult(
        faces: const [],
        frameWidth: mat.cols,
        frameHeight: mat.rows,
        tsMs: DateTime.now().millisecondsSinceEpoch,
      );
    }
    try {
      final gray = cv.cvtColor(mat, cv.COLOR_BGR2GRAY);
      final faces =
          _faceCascade!.detectMultiScale(gray, scaleFactor: 1.3, minNeighbors: 5);
      gray.dispose();

      final out = <FacePrediction>[];
      for (final r in faces) {
        final fb = FaceBox(r.x, r.y, r.width, r.height);
        final probs = _classify(mat, fb);
        final faceProbs = _renormalized(probs, _faceIndices);
        var bestI = 0;
        for (var i = 1; i < faceProbs.length; i++) {
          if (faceProbs[i] > faceProbs[bestI]) bestI = i;
        }
        final faceClass = _faceLabels[bestI];
        final faceConf = faceProbs[bestI];

        // Eyes inside face ROI (upper 60%)
        final eyeRoi = mat.region(cv.Rect(r.x, r.y, r.width, r.height));
        final grayE = cv.cvtColor(eyeRoi, cv.COLOR_BGR2GRAY);
        final eyes =
            _eyeCascade!.detectMultiScale(grayE, scaleFactor: 1.1, minNeighbors: 5);
        grayE.dispose();
        eyeRoi.dispose();

        final eyePreds = <EyePrediction>[];
        // Sort eyes by area, take 2 largest, filter to upper 60% of face
        final filtered = eyes
            .where((e) => e.y + e.height / 2 < r.height * 0.6)
            .toList()
          ..sort((a, b) => (b.width * b.height) - (a.width * a.height));
        for (final e in filtered.take(2)) {
          final eb = FaceBox(r.x + e.x, r.y + e.y, e.width, e.height);
          final ep = _classify(mat, eb);
          final ev = _renormalized(ep, _eyeIndices);
          final ei = ev[0] > ev[1] ? 0 : 1;
          eyePreds.add(EyePrediction(
            eb,
            ei == 0 ? EyeClass.closed : EyeClass.open,
            ev[ei],
          ));
        }

        out.add(FacePrediction(fb, faceClass, faceConf, eyePreds));
      }

      return DetectionResult(
        faces: out,
        frameWidth: mat.cols,
        frameHeight: mat.rows,
        tsMs: DateTime.now().millisecondsSinceEpoch,
      );
    } catch (_) {
      return DetectionResult(
        faces: const [],
        frameWidth: mat.cols,
        frameHeight: mat.rows,
        tsMs: DateTime.now().millisecondsSinceEpoch,
      );
    }
  }

  cv.Mat? _matFromCameraImage(CameraImage img) {
    try {
      // BGRA8888 (Windows / iOS when configured): single plane, 4 channels.
      if (img.format.group == ImageFormatGroup.bgra8888) {
        final bytes = img.planes[0].bytes;
        final mat = cv.Mat.fromList(
          img.height,
          img.width,
          cv.MatType.CV_8UC4,
          bytes,
        );
        return cv.cvtColor(mat, cv.COLOR_BGRA2BGR);
      }
      // YUV420 (Android default): convert via opencv_dart helper.
      if (img.format.group == ImageFormatGroup.yuv420) {
        // opencv_dart can take an interleaved YUV NV21 buffer; we build it.
        final y = img.planes[0].bytes;
        final u = img.planes[1].bytes;
        final v = img.planes[2].bytes;
        final nv21 = Uint8List(y.length + u.length + v.length);
        nv21.setAll(0, y);
        for (var i = 0, j = y.length; i < u.length; i++) {
          nv21[j++] = v[i];
          nv21[j++] = u[i];
        }
        final yuvMat = cv.Mat.fromList(
          (img.height * 3 ~/ 2),
          img.width,
          cv.MatType.CV_8UC1,
          nv21,
        );
        return cv.cvtColor(yuvMat, cv.COLOR_YUV2BGR_NV21);
      }
    } catch (_) {}
    return null;
  }

  /// Run model on a 224x224 BGR crop, return raw 6-class softmax outputs.
  Float32List _classify(cv.Mat src, FaceBox box) {
    final roi = src.region(cv.Rect(box.x, box.y, box.w, box.h));
    final resized = cv.resize(roi, (imgSize, imgSize));
    final rgb = cv.cvtColor(resized, cv.COLOR_BGR2RGB);
    roi.dispose();
    resized.dispose();

    // Build a [1,224,224,3] float32 tensor with raw uint8 values
    // (EfficientNetB0 has internal preprocessing).
    final input = Float32List(1 * imgSize * imgSize * 3);
    final raw = rgb.data; // Uint8List of HxWx3
    for (var i = 0; i < raw.length; i++) {
      input[i] = raw[i].toDouble();
    }
    rgb.dispose();

    final output = List.filled(6, 0.0).reshape([1, 6]);
    _interp!.run(input.reshape([1, imgSize, imgSize, 3]), output);
    final probs = Float32List(6);
    for (var i = 0; i < 6; i++) {
      probs[i] = (output[0][i] as num).toDouble();
    }
    return probs;
  }

  List<double> _renormalized(Float32List probs, List<int> indices) {
    final subset = indices.map((i) => probs[i]).toList();
    final sum = subset.fold<double>(0, (a, b) => a + b);
    if (sum == 0) return subset.map((_) => 0.0).toList();
    return subset.map((v) => v / sum).toList();
  }

  void dispose() {
    _interp?.close();
    _faceCascade?.dispose();
    _eyeCascade?.dispose();
  }
}

File: DrowsinessApp/src/services/detector.ts
import * as tf from '@tensorflow/tfjs';
import type { DetectionResult, EyeClass, EyePrediction, FaceBox, FaceClass, FacePrediction } from '../types';

declare const cv: any;

const MODEL_URL = '/models/efficientnet_b0/model.json';
const FACE_CASCADE_URL = '/haarcascades/haarcascade_frontalface_default.xml';
const EYE_CASCADE_URL = '/haarcascades/haarcascade_eye.xml';

const IMG_SIZE = 224;

// Class index map (from notebook):
// 0=yawn, 1=no_yawn, 2=Closed, 3=Open, 4=front, 5=down
const FACE_INDICES = [0, 1, 4, 5] as const; // softmax subset for face crop
const FACE_LABELS: FaceClass[] = ['yawn', 'no_yawn', 'front', 'down'];
const EYE_INDICES = [2, 3] as const;
const EYE_LABELS: EyeClass[] = ['Closed', 'Open'];

let model: tf.GraphModel | null = null;
let faceCascade: any = null;
let eyeCascade: any = null;
let opencvReady = false;

export function isReady(): boolean {
  return model !== null && opencvReady && faceCascade !== null && eyeCascade !== null;
}

async function waitForOpenCV(): Promise<void> {
  if (opencvReady) return;
  await new Promise<void>((resolve) => {
    const check = () => {
      if ((window as any).cv && (window as any).cv.Mat) {
        opencvReady = true;
        resolve();
      } else {
        setTimeout(check, 100);
      }
    };
    check();
  });
}

async function fetchToFS(url: string, name: string): Promise<void> {
  const resp = await fetch(url);
  const buf = await resp.arrayBuffer();
  cv.FS_createDataFile('/', name, new Uint8Array(buf), true, false, false);
}

export async function initDetector(onProgress?: (msg: string) => void): Promise<void> {
  onProgress?.('Loading TensorFlow.js…');
  await tf.ready();
  await tf.setBackend('webgl');

  onProgress?.('Loading model…');
  model = await tf.loadGraphModel(MODEL_URL);
  // warm up
  const dummy = tf.zeros([1, IMG_SIZE, IMG_SIZE, 3]);
  (model.predict(dummy) as tf.Tensor).dispose();
  dummy.dispose();

  onProgress?.('Loading OpenCV…');
  await waitForOpenCV();

  onProgress?.('Loading Haar cascades…');
  await fetchToFS(FACE_CASCADE_URL, 'face.xml');
  await fetchToFS(EYE_CASCADE_URL, 'eye.xml');
  faceCascade = new cv.CascadeClassifier();
  eyeCascade = new cv.CascadeClassifier();
  faceCascade.load('face.xml');
  eyeCascade.load('eye.xml');

  onProgress?.('Ready');
}

function softmaxOver(probs: Float32Array | number[], indices: readonly number[]): number[] {
  const subset = indices.map((i) => probs[i]);
  // already softmaxed by model — just renormalize
  const sum = subset.reduce((a, b) => a + b, 0) || 1;
  return subset.map((v) => v / sum);
}

async function classifyCrop(
  rgbaMat: any,
  box: FaceBox,
  imageData: ImageData,
): Promise<Float32Array> {
  // Crop from imageData (CPU) — faster than going through cv.Mat for tensor build.
  const { x, y, w, h } = box;
  const w2 = imageData.width;
  // Build a small offscreen canvas to crop+resize
  const off = document.createElement('canvas');
  off.width = IMG_SIZE;
  off.height = IMG_SIZE;
  const octx = off.getContext('2d')!;
  // We use a separate canvas to draw the source first, then crop via drawImage
  const srcCanvas = document.createElement('canvas');
  srcCanvas.width = w2;
  srcCanvas.height = imageData.height;
  srcCanvas.getContext('2d')!.putImageData(imageData, 0, 0);
  octx.drawImage(srcCanvas, x, y, w, h, 0, 0, IMG_SIZE, IMG_SIZE);

  return tf.tidy(() => {
    const t = tf.browser.fromPixels(off).toFloat().expandDims(0);
    const out = model!.predict(t) as tf.Tensor;
    return out.dataSync() as Float32Array;
  });
}

function detectFaces(srcMat: any): FaceBox[] {
  const gray = new cv.Mat();
  cv.cvtColor(srcMat, gray, cv.COLOR_RGBA2GRAY);
  const faces = new cv.RectVector();
  faceCascade.detectMultiScale(gray, faces, 1.3, 5);
  const boxes: FaceBox[] = [];
  for (let i = 0; i < faces.size(); i++) {
    const r = faces.get(i);
    boxes.push({ x: r.x, y: r.y, w: r.width, h: r.height });
  }
  faces.delete();
  gray.delete();
  return boxes;
}

function detectEyes(srcMat: any, face: FaceBox): FaceBox[] {
  const roi = srcMat.roi(new cv.Rect(face.x, face.y, face.w, face.h));
  const gray = new cv.Mat();
  cv.cvtColor(roi, gray, cv.COLOR_RGBA2GRAY);
  const eyes = new cv.RectVector();
  // Eyes are typically in upper half of face
  eyeCascade.detectMultiScale(gray, eyes, 1.1, 5);
  const boxes: FaceBox[] = [];
  for (let i = 0; i < eyes.size(); i++) {
    const r = eyes.get(i);
    // upper-half filter to reduce false positives (mouth/nostrils)
    if (r.y + r.height / 2 < face.h * 0.6) {
      boxes.push({
        x: face.x + r.x,
        y: face.y + r.y,
        w: r.width,
        h: r.height,
      });
    }
  }
  eyes.delete();
  gray.delete();
  roi.delete();
  // keep at most 2 largest
  return boxes.sort((a, b) => b.w * b.h - a.w * a.h).slice(0, 2);
}

export async function detectFrame(
  video: HTMLVideoElement,
  confThreshold: number,
): Promise<DetectionResult> {
  if (!isReady()) {
    return { faces: [], ts: performance.now(), faceLost: true };
  }

  const w = video.videoWidth;
  const h = video.videoHeight;
  if (w === 0 || h === 0) {
    return { faces: [], ts: performance.now(), faceLost: true };
  }

  const cap = document.createElement('canvas');
  cap.width = w;
  cap.height = h;
  const cctx = cap.getContext('2d')!;
  cctx.drawImage(video, 0, 0, w, h);
  const imageData = cctx.getImageData(0, 0, w, h);

  const src = cv.matFromImageData(imageData);
  const faceBoxes = detectFaces(src);

  const faces: FacePrediction[] = [];
  for (const fb of faceBoxes) {
    const probs = await classifyCrop(src, fb, imageData);
    const faceProbs = softmaxOver(probs, FACE_INDICES);
    let bestI = 0;
    for (let i = 1; i < faceProbs.length; i++) if (faceProbs[i] > faceProbs[bestI]) bestI = i;
    const faceClass = FACE_LABELS[bestI];
    const faceConf = faceProbs[bestI];

    const eyeBoxes = detectEyes(src, fb);
    const eyes: EyePrediction[] = [];
    for (const eb of eyeBoxes) {
      const eprobs = await classifyCrop(src, eb, imageData);
      const ep = softmaxOver(eprobs, EYE_INDICES);
      const ei = ep[0] > ep[1] ? 0 : 1;
      eyes.push({
        box: eb,
        eyeClass: EYE_LABELS[ei],
        eyeConf: ep[ei],
      });
    }

    faces.push({
      box: fb,
      faceClass,
      faceConf,
      eyes,
    });
  }

  src.delete();

  // confidence gate is applied by alertEngine, not here
  void confThreshold;

  return { faces, ts: performance.now(), faceLost: faces.length === 0 };
}
